File: app/strategy_engine/ema_crossover.py
from collections import defaultdict

import pandas as pd

from app.strategy_engine.base_strategy import (
    BaseStrategy
)

from app.signals.signal_publisher import (
    SignalPublisher
)

from app.core.logger import logger


class EMACrossoverStrategy(BaseStrategy):

    def __init__(
        self,
        fast_period: int = 5,
        slow_period: int = 10
    ):

        self.fast_period = fast_period
        self.slow_period = slow_period

        self.price_history = defaultdict(list)

    async def generate_signal(
        self,
        symbol: str,
        price: float
    ):

        history = self.price_history[symbol]

        history.append(price)

        max_history = self.slow_period + 50

        if len(history) > max_history:
            history.pop(0)

        if len(history) < self.slow_period:
            return

        prices = pd.Series(history)

        fast_ema = prices.ewm(
            span=self.fast_period
        ).mean()

        slow_ema = prices.ewm(
            span=self.slow_period
        ).mean()

        latest_fast = fast_ema.iloc[-1]
        latest_slow = slow_ema.iloc[-1]

        previous_fast = fast_ema.iloc[-2]
        previous_slow = slow_ema.iloc[-2]

        signal = None

        if (
            previous_fast <= previous_slow
            and latest_fast > latest_slow
        ):

            signal = "BUY"

        elif (
            previous_fast >= previous_slow
            and latest_fast < latest_slow
        ):

            signal = "SELL"

        if signal:

            payload = {
                "strategy": "ema_crossover",
                "symbol": symbol,
                "signal": signal,
                "price": price
            }

            logger.info(
                f"Generated Signal: {payload}"
            )

            await SignalPublisher.publish(
                payload
            )

File: backend/src/trading/strategies/emaCrossover.ts
import { BaseStrategy } from "./baseStrategy";
import { SignalPublisher } from "../signals/signalPublisher";
import { calculateEMA } from "../../utils/indicators";
import { logger } from "../../core/logger";

export class EMACrossoverStrategy extends BaseStrategy {
  private fastPeriod: number;
  private slowPeriod: number;
  private priceHistory: { [symbol: string]: number[] } = {};

  constructor(config?: { fast_period?: number; slow_period?: number }) {
    super();
    this.fastPeriod = config?.fast_period || 5;
    this.slowPeriod = config?.slow_period || 10;
  }

  async generateSignal(symbol: string, price: number): Promise<void> {
    if (!this.priceHistory[symbol]) {
      this.priceHistory[symbol] = [];
    }

    const history = this.priceHistory[symbol];
    history.push(price);

    const maxHistory = this.slowPeriod + 50;
    if (history.length > maxHistory) {
      history.shift();
    }

    if (history.length < this.slowPeriod) {
      return;
    }

    const fastEma = calculateEMA(history, this.fastPeriod);
    const slowEma = calculateEMA(history, this.slowPeriod);

    const latestFast = fastEma[fastEma.length - 1];
    const latestSlow = slowEma[slowEma.length - 1];

    const previousFast = fastEma[fastEma.length - 2];
    const previousSlow = slowEma[slowEma.length - 2];

    let signal: string | null = null;

    if (previousFast <= previousSlow && latestFast > latestSlow) {
      signal = "BUY";
    } else if (previousFast >= previousSlow && latestFast < latestSlow) {
      signal = "SELL";
    }

    if (signal) {
      const payload = {
        strategy: "ema_crossover",
        symbol,
        signal,
        price,
      };

      logger.info(`Generated Signal: ${JSON.stringify(payload)}`);
      await SignalPublisher.publish(payload);
    }
  }
}

// Register strategy
BaseStrategy.register("ema_crossover", EMACrossoverStrategy);
BaseStrategy.register("ema", EMACrossoverStrategy);
export default EMACrossoverStrategy;
